### PR TITLE
User flows: Sign in, create program, apply/join program

### PR DESCRIPTION
**Pull request by legitmaxwu**
Todo:

- [ ] Sync User table with Firebase Authentication #55 
- [x] Sync client with database changes

### CODE REVIEW

**Comment by nova-admin** [comment]
#### 📝Changed routes:


| [**/create**](https://risc-pczyanag3-novaforgood.vercel.app/create) | [**/**](https://risc-pczyanag3-novaforgood.vercel.app/) |
|:-:|:-:|
|<a href="https://risc-pczyanag3-novaforgood.vercel.app/create">Screenshot of /create</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/SNErTzjECsE97sOc-ZFGZTgFt1RjgwefRIWMJpWY7IHYXk7tmRdmBbCzMqjStFKDZLYHMrDskhtcdfzpNn0wZA.png">(view full size)</a> |<a href="https://risc-pczyanag3-novaforgood.vercel.app/">Screenshot of /</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/2UQSYUlVGsNv_uQS1QbOwPnSEJvPON5owZhkFExkNO-QjIwPPz7gnQW2CO4NONl4eYP2ngEDzwewx7unJAYrjQ.png">(view full size)</a> |



| [**/login**](https://risc-pczyanag3-novaforgood.vercel.app/login) | [**/program/\[slug\]/apply**](https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]/apply) |
|:-:|:-:|
|<a href="https://risc-pczyanag3-novaforgood.vercel.app/login">Screenshot of /login</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/mGIhXB51ACDJjLuvDWTdPFDiuKZr_CALlWkDm6nPNda7aLSl_q5qpbL5gCCoGCudqodBKvl1SW9_vfMDpZltXw.png">(view full size)</a> |<a href="https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]/apply">Screenshot of /program/[slug]/apply</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/cJfqGuDI_zI4CbFiCMBrQ6D_j9ndPqCx0SdU2--9gZLJt93KWUwB7vx8z3FSGUSKtyt4dLL-tWqePTfFJgYjsw.png">(view full size)</a> |



| [**/program/\[slug\]**](https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]) | [**/program/\[slug\]/join**](https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]/join) |
|:-:|:-:|
|<a href="https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]">Screenshot of /program/[slug]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/VVHrVJSxl8bDseZHX9WyWi6ZT3x7AaCC9EkyP-9vpynKYurB2HndtPFRccM5UjN3DRC2gjmJmWSTweC-qj6d_A.png">(view full size)</a> |<a href="https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]/join">Screenshot of /program/[slug]/join</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/HXJcJgeoS1y4DqRVt-7Cl5E7Uu-p3qnU7fIsC7Frd5XUHDKgzBc9W0_4lJz-ZVPs738uCjUOGfq-yzJ4Op74NA.png">(view full size)</a> |



And 5 other routes:
- [**/program/\[slug\]/mentees/applications**](https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]/mentees/applications)
- [**/program/\[slug\]/mentors/applications**](https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]/mentors/applications)
- [**/program/\[slug\]/mentors**](https://risc-pczyanag3-novaforgood.vercel.app/program/[slug]/mentors)
- [**/signup**](https://risc-pczyanag3-novaforgood.vercel.app/signup)
- [**/test/testmodal**](https://risc-pczyanag3-novaforgood.vercel.app/test/testmodal)

5 deleted routes:
- **/test/testtablayout**
- **/test/testtablayout/subpage1**
- **/test/testtablayout/subpage2**
- **/test/testtablayout/subpage3**
- **/test/testtablayout/subpage4**

Commit ee8d15e26d6b7f0b92edcdcf782ffe18aa346d69 (https://risc-pczyanag3-novaforgood.vercel.app).

**Review by innopoop — COMMENTED** [review]
For now, the only functional thing I caught :0

File: src/pages/program/[slug]/apply.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import Page from "../../../types/Page";
+import { useAuth } from "../../../utils/firebase/auth";
+
+const ProgramApplyPage: Page = (_) => {
+  const { loading, user } = useAuth();
+  const { currentProgram } = useCurrentProgram();
+  const [createApplication] = useCreateApplicationMutation();
+  const authorizationLevel = useAuthorizationLevel();
+
+  if (loading) return <Fragment />;
```
[the remark this answers is not in the document]
Comment (legitmaxwu) [reply src/pages/program/[slug]/apply.tsx]: Should be fixed

File: src/pages/_app.tsx [hunk 2]
```diff
@@ -24,11 +32,41 @@ const authLink = setContext(async (_, { headers, ...context }) => {
   };
 });
 
-const client = new ApolloClient({
-  link: concat(authLink, uploadLink),
+const errorLink = onError(({ graphQLErrors, networkError, response }) => {
```
Comment (kevinl120) [inline src/pages/_app.tsx:35]: Remove errorLink

File: src/utils/firebase/auth.tsx [hunk 3]
```diff
@@ -19,53 +20,45 @@ interface AuthContext {
...
+  const [user, setUser] = useState<firebase.User | null>(null);
+  const [loading, setLoading] = useState(true);
 
   const signUpWithEmail = async (email: string, password: string) => {
-    setLoading(true);
-    return firebase
-      .auth()
-      .createUserWithEmailAndPassword(email, password)
-      .finally(() => setLoading(false));
+    return firebase.auth().createUserWithEmailAndPassword(email, password);
```
Comment (kevinl120) [inline src/utils/firebase/auth.tsx:27]: On successful sign in or sign up, cache should also be cleared. (in `.then`?)

File: src/pages/_app.tsx [hunk 4]
```diff
@@ -24,11 +32,41 @@ const authLink = setContext(async (_, { headers, ...context }) => {
...
+    console.log(`[Network error]: ${networkError}`);
+  }
+});
+
+export const client = new ApolloClient({
+  link: from([authLink, errorLink, uploadLink]),
   cache: new InMemoryCache(),
 });
 
+export const getApolloClient = (
```
Comment (kevinl120) [inline src/pages/_app.tsx:59]: Add todo to see if we can do this with the same apollo client

**Review by kevinl120 — CHANGES_REQUESTED** [review]

File: src/layouts/RedirectIfNotLoggedIn.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+const RedirectIfNotLoggedIn: React.FC<RedirectIfNotLoggedInProps> = ({
+  children,
+  pathAfterLoggingIn,
+}) => {
+  const { user } = useAuth();
+  const router = useRouter();
+  return (
+    <Fragment>
+      {children}
+      <Modal isOpen={!user} onClose={() => {}}>
```
Comment (innopoop) [inline src/layouts/RedirectIfNotLoggedIn.tsx:19]: Would we just use the same function as the button for this onClose?

Comment (legitmaxwu) [reply src/layouts/RedirectIfNotLoggedIn.tsx:19]: Yeah--though this modal won't be closable so I'm passing in an empty function here.

File: src/pages/create.tsx [hunk 6]
```diff
@@ -163,9 +193,7 @@ const CreateProgramPage = () => {
                 validateProgramIdentifier(programIdentifier)
               ) {
                 setError("");
-                setLoading(true);
                 callCreateProgram();
-                setStage((prev) => prev + 1);
```
Comment (innopoop) [inline src/pages/create.tsx:168]: Are we handling stuff like if you receive an error, undo the creation of the program?

**Review by innopoop — COMMENTED** [review]
Good job max

**Review by kevinl120 — APPROVED** [review]

**Review by innopoop — APPROVED** [review]

**Pull request merged**